pandas/core/frame.py

```diff
@@ -1337,6 +1344,10 @@ def to_numpy(self, dtype=None, copy: bool = False) -> np.ndarray:
                [2, 4.5, Timestamp('2000-01-02 00:00:00')]], dtype=object)
         """
         result = np.array(self.values, dtype=dtype, copy=copy)
+
+        if na_value is not lib.no_default:
+            result[isna(result)] = na_value
```

Comment:
```
>>> df = pd.DataFrame({"A": [1, 2, None]})
>>> df.to_numpy(na_value=pd.NA)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\pandas\pandas\core\frame.py", line 1349, in to_numpy
    result[isna(result)] = na_value
TypeError: float() argument must be a string or a number, not 'NAType'
>>>
```
do we want a more user friendly message, or perhaps validate the fill_value.

Reply:
That will be difficult in general, I think (since it depends on the behaviour of numpy on what it accepts as value to set to the array). 
For this specific case, `pd.NA` is basically only possible for object dtype (so that could be checked specifically. Although it is also not the typical use case for `to_numpy`, since you typically want to get rid of pd.NA, since that doesn't work nicely with numpy).

Now, for an extension array, you already get a slightly better error message:

```
In [18]: pd.array([1, 2, None]).to_numpy(float, na_value=pd.NA)   
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-18-3c4196341619> in <module>
----> 1 pd.array([1, 2, None]).to_numpy(float, na_value=pd.NA)

~/scipy/pandas/pandas/core/arrays/masked.py in to_numpy(self, dtype, copy, na_value)
    145             ):
    146                 raise ValueError(
--> 147                     f"cannot convert to '{dtype}'-dtype NumPy array "
    148                     "with missing values. Specify an appropriate 'na_value' "
    149                     "for this dtype."

ValueError: cannot convert to '<class 'float'>'-dtype NumPy array with missing values. Specify an appropriate 'na_value' for this dtype.
```